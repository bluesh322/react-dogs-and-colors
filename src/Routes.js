import { Route, Switch, Redirect } from "react-router-dom";
import { useState } from "react";
import DogList from "./DogList";
import Dog from "./Dog";
import Color from "./Color";
import ColorFactory from "./ColorFactory";
import NewColorForm from "./NewColorForm";

const Routes = ({dogs, startColors}) => {
  const [colors, updateColors] = useState(startColors);

  const handleAdd = (newColor) => {
    updateColors(colors => ({...colors, ...newColor}));
  }
  const renderColorSelect = (props) => {
    const { color } = props.match.params;
    const value = colors[color];
    return <Color {...props} value={value} color={color} />;
  }
  return (
    <Switch>
      <Route exact path="/dogs">
        <DogList dogs={dogs}/>
      </Route>
      <Route exact path="/dogs/:name">
        <Dog dogs={dogs}/>
      </Route>
      <Route exact path="/colors">
          <ColorFactory colors={colors}/>
      </Route>
      <Route exact path="/colors/new">
        <NewColorForm updateColors={handleAdd}/>
      </Route>
      <Route exact path="/colors/:color" render={renderColorSelect}></Route>
      <Redirect to="/dogs" />
    </Switch>
  );
};



export default Routes;
