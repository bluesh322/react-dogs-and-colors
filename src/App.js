import './App.css';
import Nav from "./Nav";
import Routes from "./Routes";
import whiskey from "./static/whiskey.jpg";
import duke from "./static/duke.jpg";
import perry from "./static/perry.jpg";
import tubby from "./static/tubby.jpg"

import { BrowserRouter } from "react-router-dom"

function App({dogs, startColors}) {
  return (
    <div className="App">
      <>
      <BrowserRouter>
        <Nav/>
        <Routes dogs={dogs} startColors={startColors}/>
      </BrowserRouter>
      </>
    </div>
  );
}

App.defaultProps = {
  dogs: [
    {
      name: "Whiskey",
      age: 5,
      src: whiskey,
      facts: [
        "Whiskey loves eating popcorn.",
        "Whiskey is a terrible guard dog.",
        "Whiskey wants to cuddle with you!",
      ],
    },
    {
      name: "Duke",
      age: 3,
      src: duke,
      facts: [
        "Duke believes that ball is life.",
        "Duke likes snow.",
        "Duke enjoys pawing other dogs.",
      ],
    },
    {
      name: "Perry",
      age: 4,
      src: perry,
      facts: [
        "Perry loves all humans.",
        "Perry demolishes all snacks.",
        "Perry hates the rain.",
      ],
    },
    {
      name: "Tubby",
      age: 4,
      src: tubby,
      facts: [
        "Tubby is really stupid.",
        "Tubby does not like walks.",
        "Angelina used to hate Tubby, but claims not to anymore.",
      ],
    },
  ],
  startColors:
      {
          red: "#FF0000",
          blue: "#0000FF",      
          green: "#00FF00",
      }
};

export default App;
