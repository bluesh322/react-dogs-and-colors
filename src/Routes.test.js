import React from "react";
import { MemoryRouter } from "react-router-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import Routes from "./Routes";
import App from "./App";

const { dogs, startColors } = App.defaultProps;

test("render all dogs", () => {
  const { getByText } = render(
    <MemoryRouter initialEntries={["/dogs"]}>
      <Routes dogs={dogs} />
    </MemoryRouter>
  );

  const dogNames = dogs.map((d) => d.name);
  for (const name of dogNames) {
    const linkElements = getByText(new RegExp(name, "i"));
    expect(linkElements).toBeInTheDocument();
  }
});

test("renders only Whiskey's info", () => {
  const { getByText, queryByText } = render(
    <MemoryRouter initialEntries={["/dogs/whiskey"]}>
      <Routes dogs={dogs} />
    </MemoryRouter>
  );

  const whiskey = dogs.find((d) => d.name === "Whiskey");
  const duke = dogs.find((d) => d.name === "Duke");

  const liElement = getByText(new RegExp(whiskey.facts[0], "i"));
  expect(liElement).toBeInTheDocument();

  expect(screen.queryByText(new RegExp(duke.facts[0], "i"))).toBeNull();
  expect(queryByText(new RegExp(duke.facts[0], "i"))).not.toBeInTheDocument();
});

test("renders home on bad route", () => {
  const { getByText } = render(
    <MemoryRouter initialEntries={["/bad_route"]}>
      <Routes dogs={dogs} />
    </MemoryRouter>
  );

  const dogNames = dogs.map((d) => d.name);
  for (const name of dogNames) {
    const linkElement = getByText(new RegExp(name, "i"));
    expect(linkElement).toBeInTheDocument();
  }
});

test("render all colors", () => {
  const { getByText } = render(
    <MemoryRouter initialEntries={["/colors"]}>
      <Routes startColors={startColors} />
    </MemoryRouter>
  );

  const colorNames = Object.keys(startColors);
  for (const name of colorNames) {
    const linkElements = getByText(new RegExp(name, "i"));
    expect(linkElements).toBeInTheDocument();
  }
});

test("render red color", () => {
  //I think this should be a snapshot test?
  const { asFragment } = render(
    <MemoryRouter initialEntries={["/colors/red"]}>
      <Routes startColors={startColors} />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();

  // const red = "backgroundColor: rgb(255, 0, 0);"
  // const green = "backgroundColor: rgb(0, 255, 0);"

  // const divEl = queryByText(new RegExp(red, "i"));
  // expect(divEl).toBeInTheDocument();
});

test("should add a new color link", () => {
  const { getByText, queryByText, getByLabelText } = render(
    <MemoryRouter initialEntries={["/colors/new"]}>
      <Routes startColors={startColors} />
    </MemoryRouter>
  );

  const nameInput = getByLabelText("name");
  const colorInput = getByLabelText("value");
  const btn = queryByText("Add Color");

  fireEvent.change(nameInput, {target: {value: 'black'}})
  fireEvent.change(colorInput, {target: {value: '#000000'}})
  fireEvent.click(btn);

  //redirect to '/colors' with black as a new link
  const linkElement = getByText(new RegExp('black', "i"));
  expect(linkElement).toBeInTheDocument();

  const blackLink = getByText('black');
  fireEvent.click(blackLink);

  //can I make a snapshot of a black square and compare it here?

});
