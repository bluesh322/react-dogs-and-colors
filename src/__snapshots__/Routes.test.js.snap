// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render red color 1`] = `
<DocumentFragment>
  <div
    style="background-color: rgb(255, 0, 0); width: 400px; height: 400px;"
  >
    <p>
      Here's your color
    </p>
    <a
      href="/colors"
    >
      Go back
    </a>
  </div>
</DocumentFragment>
`;
